src/utils/async.ts

```diff
@@ -48,3 +50,43 @@ export async function asyncIterableToArray<T>(
   const iter = source[Symbol.asyncIterator]();
   return asyncIteratorToArray(iter);
 }
+
+export function mapEitherAsyncIterator<E, T, V>(
+  iter: AsyncIterator<Either<E, T>, Either<E, T>>,
+  f: (t: T) => V
+): AsyncIterator<V> {
+  return {
+    next: async () => {
+      while (true) {
+        const { done, value } = await iter.next();
+        if (done) {
+          return { done, value: undefined };
+        }
+        if (isRight(value)) {
+          return { done, value: f(value.value) };
+        }
+      }
+    }
+  };
+}
+
+export function flattenAsyncIterator<T>(
+  iter: AsyncIterator<ReadonlyArray<T>>
+): AsyncIterator<T> {
+  // tslint:disable-next-line: no-let
+  let index = 0;
+  // tslint:disable-next-line: no-let
+  let flattenArray: ReadonlyArray<T> = [];
+  return {
+    next: async () => {
+      if (flattenArray.length === index) {
+        const { done, value } = await iter.next();
+        if (done) {
+          return { done, value: undefined };
+        }
+        flattenArray = [...flattenArray, ...value];
+      }
+      return { done: false, value: flattenArray[index++] };
```

Comment:
I believe you should check again whether `index < flattenArray.length` here as `value` returned by `next()` could be an empty array

Reply:
Fixed in 57b70cb